rust/repositories.bzl

```diff
@@ -303,8 +306,22 @@ def _rust_toolchain_repository_proxy_impl(ctx):
             exec_triple = ctx.attr.exec_triple,
             parent_workspace_name = ctx.attr.parent_workspace_name,
             target_triple = target_triple,
+            toolchain_type = "toolchain",
         ))
 
+    BUILD_components.append(BUILD_for_toolchain(
+        name = "{toolchain_prefix}_wasm32-unknown-unknown".format(
+            toolchain_prefix = ctx.attr.toolchain_name_prefix,
+        ),
+        exec_triple = ctx.attr.exec_triple,
+        parent_workspace_name = ctx.attr.parent_workspace_name,
+        # Technically we should say it's compatible with wasm32-unknown-unknown but in practice we
+        # want to be able to produce wasm output when targeting the host computer (e.g. for building
+        # a docker image). Thus we make this the exec triple.
+        target_triple = ctx.attr.exec_triple,
```

Comment:
I see you have a pretty big comment here, but I'll ask anyway.

This defies my original intent for the target_triple field. My hope was that "wasm32-unknown-unknown" could be specified as a target triple in the way of other targets, and I guess, that it would never make sense to compile for host. You mention that there is a docker usecase for doing that, can you elaborate a bit?

Reply:
Yes this is a good point and I mentioned it might be the controversial part of this PR in the initial request. Basically the use-case is that we have a server binary written in rust (we can call it `//server` for example) and some client code written in TypeScript/JavaScript (let's call the bundle of compiled .js and assets `//web`). Then we have a docker image where we package it all together and ship it to production (e.g. `//:deployment_image` which depends on `//server` and `//web`). The problem is that //server needs to target something like x86_64-linux-gnu while //web needs to target wasm32-unknown-unknown. Thus for now, you wouldn't be able to build //:deployment_image because bazel only supports a single target architecture at a time (see https://stackoverflow.com/questions/51423896/bazel-build-multiple-target-architectures-with-one-invocation). You'd also have trouble running all the tests at once (`bazel test //...`) for similar reasons.

The way we solve it here is that each rule produces its normal output but also a `%name%.wasm` and that .wasm file can be built because it technically targets the host architecture.

Still I agree this is a bit messy so I'll leave it up to you whether this is a valid use-case. In the worst-case scenario we can just call bazel twice with the different constraints and then introduce a mini-build-system on top of bazel to do the final image assembly. So I'd be willing to change this PR so it acts a bit more like you'd expect where you can specify the constraints to pick different toolchains.

Reply:
I did some more digging and found https://docs.google.com/document/d/1vc8v-kXjvgZOdQdnxPTaV0rrLxtP2XwnD2tAZlYJOqw/edit#heading=h.pn8nbhno42p5 which says it should allow custom transitions between constraints. So I think you're right and we can just remove all the additional outputs and implement wasm mostly normally.

Reply:
Ah I understand the problem. The output target is not wasm32-unknown-unknown, but rather the server platform. The Rust code involved in the output artifact is closer to "data" for the final target than to an executable for the configured Bazel platform.

This seems to be a specific instance of the general problem of building an "executable" that serves "executable". This is clearly an underserved use case for Bazel. My cursory research on the subject indicates that this is typically handled with a "wrapper" build rule over the standard build rules, that uses Bazel's experimental transition feature[1]. I think this is what you linked in your latest comment

This goes pretty far outside my current knowledge of Bazel platform configuration support but I agree with your latest assessment (which agrees with my initial suspicion) -- we should ideally be able to support that use case using the existing toolchain options.

[1] https://docs.bazel.build/versions/master/skylark/lib/transition.html

Reply:
I pushed some changes to use Bazel's transition feature. Bazel 0.29.0 was released yesterday with the fix for https://github.com/bazelbuild/bazel/issues/8963 so it should work now. The only thing I'm still trying to figure out is why RBE image fails on the `matrix_dylib_test`.